File: disposable.js
'use strict';

var domainMap = new Set();
var arr = require('./domains')

for (var i = 0; i < arr.length; ++i)
    domainMap.add([arr[i]])

module.exports = {
    validate: function(domainOrEmail, callback) {
        var domain = domainOrEmail.split('@').pop()
        var isValid = !domainMap.has(domain)

        if (!callback) {
            return isValid
        }
        callback(null, isValid)
    }
}

File: disposable.go
package disposable

import (
	"io/ioutil"
	"strings"
)

const inFile = "domains.txt"

var domainMap = map[string]struct{}{}

func init() {
	dat, err := ioutil.ReadFile(inFile)
	if err != nil {
		panic("Missing (disposable email) domains file")
	}
	for _, domain := range strings.Split(string(dat), "\n") {
		domainMap[domain] = struct{}{}
	}
}

// Domain tests whether a string is among the known set of disposable mailboxes
func Domain(d string) bool {
	if _, ok := domainMap[d]; ok {
		return true
	}
	return false
}

// DomainMap returns all known domains in form of a map (for O(1)-ish search)
func DomainMap() map[string]struct{} {
	return domainMap
}
